plugins/h3/h3.cpp

```diff
@@ -0,0 +1,247 @@
+{
+    int finger = 0;
+    for (int i = 0; i < maxBuffLen; ++i)
+    {
+        if (buff[i] == 0)
+            continue;
+
+        if (i > finger)
+        {
+            buff[finger] = buff[i];
+            buff[i] = 0;
+        }
+        ++finger;
+    }
+    __isAllResult = false;
+    __lenResult = finger * sizeof(H3Index);
+    __result = buff;
+}
+
+namespace h3
+{
+
+//--------------------------------------------------------------------------------
+//                           ECL SERVICE ENTRYPOINTS
+//--------------------------------------------------------------------------------
+
+//  Indexing Functions ---
+ECL_H3_API unsigned __int64 ECL_H3_CALL index(ICodeContext *_ctx, double lat, double lon, uint32_t resolution)
+{
+    GeoCoord location;
+    location.lat = ::degsToRads(lat);
+    location.lon = ::degsToRads(lon);
+    return ::geoToH3(&location, resolution);
+}
+
+ECL_H3_API void ECL_H3_CALL center(ICodeContext *_ctx, size32_t &__lenResult, void *&__result, unsigned __int64 index)
+{
+    GeoCoord location;
+    ::h3ToGeo(index, &location);
+    __lenResult = 2 * sizeof(double);
+    __result = rtlMalloc(__lenResult);
+    double *cur = static_cast<double *>(__result);
+    *cur++ = ::radsToDegs(location.lat);
+    *cur = ::radsToDegs(location.lon);
+}
+
+ECL_H3_API void ECL_H3_CALL boundary(ICodeContext *_ctx, size32_t &__lenResult, void *&__result, unsigned __int64 index)
+{
+    GeoBoundary boundary;
+    ::h3ToGeoBoundary(index, &boundary);
+    __lenResult = boundary.numVerts * 2 * sizeof(double);
+    __result = rtlMalloc(__lenResult);
+
+    double *cur = static_cast<double *>(__result);
+    for (int v = 0; v < boundary.numVerts; v++)
+    {
+        *cur++ = ::radsToDegs(boundary.verts[v].lat);
+        *cur++ = ::radsToDegs(boundary.verts[v].lon);
+    }
+}
+
+//  Introspection ---
+ECL_H3_API uint32_t ECL_H3_CALL resolution(ICodeContext *_ctx, unsigned __int64 index)
+{
+    return ::h3GetResolution(index);
+}
+
+ECL_H3_API uint32_t ECL_H3_CALL baseCell(ICodeContext *_ctx, unsigned __int64 index)
+{
+    return ::h3GetBaseCell(index);
+}
+
+ECL_H3_API void ECL_H3_CALL toString(ICodeContext *_ctx, size32_t &lenVarStr, char *&varStr, unsigned __int64 index)
+{
+    varStr = static_cast<char *>(rtlMalloc(H3_INDEX_MINSTRLEN));
+    ::h3ToString(index, varStr, H3_INDEX_MINSTRLEN);
+    lenVarStr = strlen(varStr);
+}
+
+ECL_H3_API unsigned __int64 ECL_H3_CALL fromString(ICodeContext *_ctx, const char *strIdx)
+{
+    return ::stringToH3(strIdx);
+}
+
+ECL_H3_API bool ECL_H3_CALL isValid(ICodeContext *_ctx, unsigned __int64 index)
+{
+    return ::h3IsValid(index) != 0;
+}
+
+//  Traversal  ---
+ECL_H3_API void ECL_H3_CALL kRing(ICodeContext *_ctx, bool &__isAllResult, size32_t &__lenResult, void *&__result, unsigned __int64 index, int32_t k)
+{
+    int maxBuff = ::maxKringSize(k);
+    H3Index *buff = static_cast<H3Index *>(calloc(maxBuff, sizeof(H3Index)));
```

Comment:
theoretically, if this memory is being returned to the query it should be using rtlMalloc instead of calloc.  In practice I don't think it will cause a problem, other plugins do the same (e.g., returning buffer.detach)

Reply:
do we have a rtlCalloc?  otherwise I will need to pre clear it.